### PR TITLE
module_adapter: always propagate prepare() to module

### PR DESCRIPTION
**Pull request by johnylin76**
At present, module_adapter only propagates prepare() to the module for the first time. For modules, the specific reset() procedure only restores the initial stage after prepared, init_process() is applied to start module processing. Therefore, module_adapter blocks all the following prepare() calls to module.

However, it leads to problems that modules won't be notified for the stream parameter change once it happens afterwards, since prepare() is blocked by module_adapter. It actually breaks the ALSA conformance test in Google auto-lab, as the test
automation will run with S16, S24, S32 sample format successively.

This commit removes the logic of blocking prepare() in module_adapter which makes sure that the stream parameter change can be informed to modules. Accordingly, modules should be responsible for handling stream parameter in an adaptive manner.

### CODE REVIEW

File: src/audio/module_adapter/module/cadence.c [hunk 1]
```diff
@@ -490,6 +490,15 @@ static int cadence_codec_prepare(struct processing_module *mod)
...
+	if (mod->priv.state >= MODULE_IDLE) {
+		/* TODO: handle the case that stream_params was changed afterwards (different from
+		 *       the one applied on module codec initialization).
+		 */
+		comp_dbg(dev, "cadence_codec_prepare(): skipped by module under state %d",
+			 mod->priv.state);
+		return 0;
+	}
+
 	comp_dbg(dev, "cadence_codec_prepare() start");
```
Comment (dbaluta) [inline src/audio/module_adapter/module/cadence.c:502]: I think we should merge this patch once the `TODO` is implemented. 

First observation is that we need to add the `if` above after the call to `cadence_codec_post_init`. For example if the codec id changes we won't see that because the codec is resolved inside `cadence_codec_post_init`.

Comment (johnylin76) [reply src/audio/module_adapter/module/cadence.c:502]: I am fine with that, although this patch should cause no behavior change as it is.
Who should be in charge of the implementation?

Comment (dbaluta) [reply src/audio/module_adapter/module/cadence.c:502]: @johnylin76 I assume since you sent this patch then there is an use case you want to support, right? 

Cc: @cujomalainey

Comment (johnylin76) [reply src/audio/module_adapter/module/cadence.c:502]: @dbaluta I don't see it that way.

This patch won't change the current behavior. The module specific prepare() will be called while in MODULE_IDLE state, and early returned by all existing modules without doing anything.

This patch is to provide the flow for modules to handle re-configuration for stream params change, which is not handled (even not informed) now. There are multiple ways to handle and it's better to be decided and implemented by the module owners, since they have the deeper knowledge for the processing in modules.

Comment (dbaluta) [reply src/audio/module_adapter/module/cadence.c:502]: Ok, I agree. we can merge this now.

File: src/audio/module_adapter/module/generic.c [hunk 2]
```diff
@@ -197,15 +197,13 @@ int module_prepare(struct processing_module *mod)
...
-		return 0;
 	if (mod->priv.state < MODULE_INITIALIZED)
 		return -EPERM;
 
 	ret = md->ops->prepare(mod);
 	if (ret) {
-		comp_err(dev, "module_prepare() error %d: module specific prepare failed, comp_id %d",
-			 ret, dev_comp_id(dev));
+		comp_err(dev, "module_prepare() error %d: module specific prepare failed, comp_id %din state %d",
+			 ret, dev_comp_id(dev), md->state);
```
Comment (dbaluta) [inline src/audio/module_adapter/module/generic.c:206]: This change should go in a different patch as it is not strictly related to current patch.

**Comment by dbaluta** [comment]
@ranj063 can you also please have a look?

File: src/audio/module_adapter/module/generic.c (hunk 2 above)
Comment (johnylin76) [reply src/audio/module_adapter/module/generic.c:206]: Before this patch, module specific prepare() is called only in MODULE_INITIALIZED state.
After this patch, it will be called in MODULE_INITIALIZED and MODULE_IDLE states. So it becomes crucial to show the state in error message.

Comment (dbaluta) [reply src/audio/module_adapter/module/generic.c:206]: @johnylin76 yes, but this can be done in a separate patch.

**Review by lgirdwood — COMMENTED** [review]
@johnylin76 this is a good catch, I think the original intention of the author was to follow the flow that was being used by the processing module from Windows (but this may be exact copy atm).

@jxstelter can you give advice here. When is ```prepare()``` allowed with Windows modules (in which module states can it be called) ? In Linux ALSA can call prepare() during stream recovery or a stream params renegotiation.

**Comment by johnylin76** [comment]
@lgirdwood thanks for the information, my previous thought about the design intention is to reduce the repetitive memory allocation for modules through reset->prepare routines.
If so, another design idea for re-configuration is to increase the flexibility for module implementations to choose either of the following state flow:
1. as it is: module specific reset() won't de-allocate memories, no further prepare() is needed.
2. as component API: module specific reset() de-allocates memories, prepare() is needed every time for memory allocation and stream param update.

In fact it is even easier to make the present code be compatible to both state flows, we just need to revise the definition for return code of [module specific reset()](https://github.com/thesofproject/sof/blob/main/src/audio/module_adapter/module/generic.c#L272). For example, return 0 for case#1, 1 for case#2, negative values for errors. The generic code sets state to MODULE_IDLE for case#1, MODULE_INITIALIZED for case#2. And we're all done.

This design has another benefit: memory usage. modules in case#2 free the memory when they don't need. It makes more senses for modules like audio enhancement, which has been placed on every playback pipeline but typically only one of them is active.

**Comment by jxstelter** [comment]
> @johnylin76 this is a good catch, I think the original intention of the author was to follow the flow that was being used by the processing module from Windows (but this may not be exact copy atm).
> 
> @jxstelter can you give advice here. When is `prepare()` allowed with Windows modules (in which module states can it be called) ? In Linux ALSA can call prepare() during stream recovery or a stream params renegotiation.

Actually there is no 'prepare()' call in ProcessingModuleInterface in Windows. So all memory allocations are done during 'init()'.  I believe 'reset()' is not freeing any memory, it just bring module in its well-known initial state. So stream recovery is possible. In case of stream format change, it require pipeline re-creation.

File: src/audio/module_adapter/module/cadence.c (hunk 1 above)
Comment (cujomalainey) [reply src/audio/module_adapter/module/cadence.c:502]: @dbaluta to clear things up, the issue is if we change pipe params, and dsp does not suspend between params calls, the result is the codec does not get new information and is misconfigured at playback and results in flow rate errors. We are not sure why this is designed this way, but we are pretty sure it is incorrect.

**Comment by johnylin76** [comment]
@jxstelter I think there are some differences between ProcessingModuleInterface and SOF Component API (of Linux ALSA usage). ALSA pipelines are basically all static, components (modules) are allocated on .create() from the start. For processing a upcoming stream, the parameter/format is set through .params(), then runtime data buffers are allocated on .prepare(), and freed on .reset() in the end of stream.

In my point of view, the major problem of present module_adapter design is that it omits the propagation of .params() and .prepare() to the module inside. It seems unnecessary for ProcessingModuleInterface wrapping, but is critical for modules designed as ALSA usage.

**Comment by lyakh** [comment]
does #6184 help?

**Comment by johnylin76** [comment]
#6184 also points out the similar issue which affects volume module for the present module_adapter design.

@ranj063 @lgirdwood could you help to provide comments about this?
Just as I mentioned [in this comment](https://github.com/thesofproject/sof/pull/6166#issuecomment-1225380428), the design needs to be revised to be compatible for both Windows modules (applied ProcessingModuleInterface) and ALSA modules.

**Comment by johnylin76** [comment]
> does #6184 help?

The fix of #6184 may break other modules with the unexpected prepare() call.

File: src/audio/module_adapter/module/volume/volume.c [hunk 3]
```diff
@@ -1182,6 +1182,15 @@ static int volume_prepare(struct processing_module *mod)
 	int ret;
 	int i;
 
+	if (mod->priv.state >= MODULE_IDLE) {
```
Comment (johnylin76) [inline src/audio/module_adapter/module/volume/volume.c:1185]: @lyakh I think this commit can also fix #6184 by removing this early-return logic, while keeping other modules work as same behavior.

**Comment by lgirdwood** [comment]
> #6184 also points out the similar issue which affects volume module for the present module_adapter design.
> 
> @ranj063 @lgirdwood could you help to provide comments about this? Just as I mentioned [in this comment](https://github.com/thesofproject/sof/pull/6166#issuecomment-1225380428), the design needs to be revised to be compatible for both Windows modules (applied ProcessingModuleInterface) and ALSA modules.

Agree, we need to be compatible with both Windows and Chrome. 

I would like to avoid the extra logic in the module as module binaries tested on Windows could also be loaded on Chrome/Linux. Can we push this logic up the stack into the module adapter ? 

Fwiw, @ranj063 is back next week and can help here as the author of the code.

File: src/audio/module_adapter/module/volume/volume.c (hunk 3 above)
Comment (lyakh) [reply src/audio/module_adapter/module/volume/volume.c:1185]: you mean as is this PR will not fix #6034 and to make it fix it this change to volume.c should be removed? Then probably this should be done? In its present form this PR shouldn't change behaviour of any existing modules, right? `.prepare()` will now be propagated to the modules, but then all in-tree modules will immediately return since after `.reset()` the module will be in `MODULE_IDLE` state, is that correct?

**Comment by mwasko** [comment]
> > #6184 also points out the similar issue which affects volume module for the present module_adapter design.
> > @ranj063 @lgirdwood could you help to provide comments about this? Just as I mentioned [in this comment](https://github.com/thesofproject/sof/pull/6166#issuecomment-1225380428), the design needs to be revised to be compatible for both Windows modules (applied ProcessingModuleInterface) and ALSA modules.
> 
> Agree, we need to be compatible with both Windows and Chrome.
> 
> I would like to avoid the extra logic in the module as module binaries tested on Windows could also be loaded on Chrome/Linux. Can we push this logic up the stack into the module adapter ?
> 
> Fwiw, @ranj063 is back next week and can help here as the author of the code.

We might also need to take a look at a SOF driver and how the ALSA pcm ops are implemented. The IPC4 support is a good opportunity to adjust the pcm handling and align with new FW Module API. Overall the Windows and Chrome compatibility is expected from IPC4 and onwards.

File: src/audio/module_adapter/module/volume/volume.c (hunk 3 above)
Comment (johnylin76) [reply src/audio/module_adapter/module/volume/volume.c:1185]: Yes, exactly. And I stated an example of the bad influence to other modules if we only modify the generic code to make .prepare() propagated to modules every time. (https://github.com/thesofproject/sof/pull/6184#issuecomment-1229174345).

**Comment by johnylin76** [comment]
>

**Comment by johnylin76** [comment]
Sorry for closing by mistake...

Regarding the suggestion from @lgirdwood :
>  would like to avoid the extra logic in the module as module binaries tested on Windows could also be loaded on Chrome/Linux. Can we push this logic up the stack into the module adapter ?

I think the solution https://github.com/thesofproject/sof/pull/6166#issuecomment-1224142678 should be more suitable. I created #6216 for implementation which also includes the volume fix.
@lyakh @lgirdwood does it look better to you?

**Comment by lgirdwood** [comment]
> I think the solution [#6166 (comment)](https://github.com/thesofproject/sof/pull/6166#issuecomment-1224142678) should be more suitable. I created #6216 for implementation which also includes the volume fix.
> @lyakh @lgirdwood does #6216 look better to you?

I will leave the decision to @ranj063 as she knows the module API better than me. My acceptance criteria is that we can run the same binary modules from Windows without modification.

**Comment by johnylin76** [comment]
Fixed by #6230 instead
